oak_containers_orchestrator/src/container_runtime.rs

```diff
@@ -0,0 +1,114 @@
+//
+// Copyright 2023 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// Representation of a minimal OCI filesystem bundle config, including just the required fields
+/// Ref: <https://github.com/opencontainers/runtime-spec/blob/467fd17d4f2a987fa00051ce44b69bf9620e2ee6/config.md>
+#[derive(serde::Deserialize)]
+struct OciFilesystemBundleConfig {
+    process: OciFilesystemBundleConfigProcess,
+    root: OciFilesystemBundleConfigRoot,
+}
+
+#[derive(serde::Deserialize)]
+struct OciFilesystemBundleConfigProcess {
+    user: OciFilesystemBundleConfigProcessUser,
+    args: Vec<String>,
+}
+
+#[derive(serde::Deserialize)]
+struct OciFilesystemBundleConfigProcessUser {
+    uid: u32,
+    gid: u32,
+}
+
+#[derive(serde::Deserialize)]
+struct OciFilesystemBundleConfigRoot {
+    path: std::path::PathBuf,
+}
+
+async fn run_command_and_log_output(
+    command: &mut tokio::process::Command,
+) -> Result<(), Box<std::io::Error>> {
+    let output = command.output().await;
+    log::info!("{:?}: {:?}", command, output);
+    Ok(())
+}
+
+// Directory at which the container OCI filesystem bundle will be unpacked
+const CONTAINER_DIR: &str = "/oak_container";
+
+pub async fn run(container_bundle: &[u8]) -> Result<(), anyhow::Error> {
+    tokio::fs::create_dir(CONTAINER_DIR).await?;
+    tar::Archive::new(container_bundle).unpack(CONTAINER_DIR)?;
+
+    let oci_filesystem_bundle_config: OciFilesystemBundleConfig = {
+        let file_path = {
+            let mut base = std::path::PathBuf::from(CONTAINER_DIR);
+            base.push("config.json");
+            base
+        };
+        let oci_filesystem_bundle_config_file = tokio::fs::read_to_string(file_path).await?;
+
+        serde_json::from_str(&oci_filesystem_bundle_config_file)?
+    };
+
+    // mount host vsock device inside the container directory to allow the
+    // trusted application to communicate with the outside
+    {
+        run_command_and_log_output(
+            tokio::process::Command::new("rm")
+                .current_dir(CONTAINER_DIR)
+                .arg("-rf")
+                .arg("dev"),
+        )
+        .await?;
+        run_command_and_log_output(
+            tokio::process::Command::new("mkdir")
+                .current_dir(CONTAINER_DIR)
+                .arg("-p")
+                .arg("dev/vsock"),
```

Comment:
Don't we need other devices to be avialable here as well?

Also, should we mount procfs and sysfs?

Reply:
For now I went the minimal device needed for the example. 

We could mount all the ones you suggested by default, or we could make containers [list them required mountpoints](https://github.com/opencontainers/runtime-spec/blob/v0.2.0/config.md#mount-points) the config.json of the filesystem bundle we load. Which do you think is better?

Reply:
I'm pretty sure we'll need to mount `/proc`, sysfs may not be that important.

Does the average container (eg anything you export from Docker) list the required mountpoints in a reasonable way? If yes, then indeed relying on that seems like a good idea.

Reply:
> Does the average container (eg anything you export from Docker) list the required mountpoints in a reasonable way?

I don't think it does, container runtimes seem to be using some magic to eg expose a filtered version of /proc, and the same for some devices (eg /dev/mqueue) and sys/fs. 

Probably the best is to yolo and just expose all for now, even I don't love it.

Reply:
I am in favour of just passing everything through. We don't plan on isolating containers from each other (only one containers per VM) or relying on the containers as a security boundary, so I think making all the expected pseudo-filesystems available without filtering is fine.